src/qibolab/compilers/compiler.py

```diff
@@ -0,0 +1,171 @@
+            the given gate.
+        virtual_z_phases (dict): Dictionary mapping qubits to virtual Z-phases induced by the gate.
+
+    See ``qibolab.compilers.default`` for an example of a compiler implementation.
+    """
+
+    rules: dict = field(default_factory=dict)
+    """Map from gates to compilation rules."""
+
+    @classmethod
+    def default(cls):
+        return cls(
+            {
+                gates.I: identity_rule,
+                gates.Z: z_rule,
+                gates.RZ: rz_rule,
+                gates.U3: u3_rule,
+                gates.CZ: cz_rule,
+                gates.M: measurement_rule,
+            }
+        )
+
+    def __setitem__(self, key, rule):
+        """Sets a new rule to the compiler.
+
+        If a rule already exists for the gate, it will be overwritten.
+        """
+        self.rules[key] = rule
+
+    def __getitem__(self, item):
+        """Get an existing rule for a given gate."""
+        if item not in self.rules:
+            raise_error(KeyError, f"Compiler rule not available for {item}.")
+        return self.rules[item]
+
+    def register(self, gate_cls):
+        """Decorator for registering a function as a rule in the compiler.
+
+        Using this decorator is optional. Alternatively the user can set the rules directly
+        via ``__setitem__`.
+
+        Args:
+            gate_cls: Qibo gate object that the rule will be assigned to.
+        """
+
+        def inner(func):
+            self[gate_cls] = func
+            return func
+
+        return inner
+
+    def remove(self, item):
+        """Remove rule for the given gate."""
+        if item not in self.rules:
+            raise_error(KeyError, f"Cannot remove {item} from compiler because it does not exist.")
+        self.rules.pop(item)
+
+    def __call__(self, circuit, platform):
+        """Transforms a circuit to pulse sequence.
+
+        Args:
+            circuit (qibo.models.Circuit): Qibo circuit that respects the platform's
+                connectivity and native gates.
+            platform (qibolab.platforms.abstract.AbstractPlatform): Platform used
+                to load the native pulse representations.
+
+        Returns:
+            sequence (qibolab.pulses.PulseSequence): Pulse sequence that implements the circuit.
+        """
+        sequence = PulseSequence()
+        # FIXME: This will not work with qubits that have string names
+        # TODO: Implement a mapping between circuit qubit ids and platform ``Qubit``s
+        virtual_z_phases = defaultdict(int)
+
+        measurement_map = MeasurementMap()
+        # keep track of gates that were already added to avoid adding them twice
+        already_processed = set()
+        # process circuit gates
+        for moment in circuit.queue.moments:
+            moment_start = sequence.finish
+            for gate in moment:
+                if gate is not None and gate not in already_processed:
```

Comment:
If I understood this `already_processed` is a way to process each gate only once, even though they are distributed on several *moments*, and possibly repeated.

Instead, a possible simplification would be to sort them ahead of time, and directly iterate the sorted and deduplicated sequence.

In particular, this operation can be done well in a separate function before the iteration, denesting a further bit, and decoupling the operations (while the `already_processed` strategy requires some operations to be performed both outside and inside the loop).
The function would only depend on `circuit.queue.moments`:

```suggestion
        gates = gates_history(circuit.queue.moments)
        
        for gate in gates:
            if gate is None:
                continue

            moment_start = sequence.finish
```

(unfortunately you would compute more often `moment_start`, but if you want to avoid just return from `gates_history` a tuple with it as well - and please find a better name than `gates_history`...)

Reply:
I am not sure how I would define `gates_history` in this case, other than naively moving the current approach to an iterator
```py
def gates_history(moments):
    already_processed = set()
    for moment in moments:
        if gate is not None and gate not in already_processed:
            yield gate
        already_processed.add(gate)
```
however even this does not seem possible because `moment_start` depends on the `sequence` which is not available in the iterator (or the first loop). So I am not sure if we can actually denest this while keeping the current behavior.